LSPS3/README.md

```diff
@@ -0,0 +1,317 @@
+
+> **Rationale** This allows future versions of this spec to add new
+> fields to this response, without causing problems for clients that
+> only support the current version.
+
+### 2.  Request JIT Channel
+
+The client constructs a request body for a `POST /lsp/jitchannel`
+request, depending on their need.
+
+Example `POST /lsp/jitchannel` request:
+
+```JSON
+{
+    "version": 1,
+    "opening_fee": "546000",
+    "opening_fee_valid_until": "2023-02-23T08:47:30.511Z",
+    "opening_fee_promise": "abcdefghijklmnopqrstuvwxyz",
+    "payment_size": "42000",
+    "client_node_id": "03d4e028a0d4a90868ec202ab684fb0085779defea9ca7553e06146557631eec20"
+}
+```
+
+`version` is the version of this spec that will be used for this
+interaction.
+
+`opening_fee`, `opening_fee_valid_until`, and `opening_fee_promise`
+are the parameters acquired from the previous step.
+Clients MUST copy them verbatim from some previous
+`GET /lsp/jitchannel` response.
+LSPs SHOULD check that the `opening_fee_promise` does in fact
+prove that it previously promised the specified `opening_fee` and
+`opening_fee_valid_until`.
+LSPs SHOULD check that the `opening_fee_valid_until` is not a
+past datetime.
+
+`payment_size` is the amount that the client wants to receive.
+
+The client MUST ensure that `payment_size` is within the previous
+`min_payment_size` and `max_payment_size` parameters from the LSP.
+The LSP MUST validate that the `payment_size` is within the previous
+`min_payment_size` and `max_payment_size` parameters from the LSP.
+
+The client MAY use a "subtractive" computation to get the
+`payment_size`.
+For example, the client may want to charge some payer an exact
+amount of 1,000,000 millisatoshis, but the LSP indicated an
+`opening_fee` of 546,000 millisatoshis.
+In that case, the client should specify 454,000 millisatoshis for
+`payment_size`, so that the invoice charges 1,000,000 millisatoshis
+total.
+
+`client_node_id` is the node ID of the client.
+
+The LSP can then respond with a response object or an error
+object.
+The client can differentiate between these two objects by checking
+for an `error` field, which exists and is the Boolean value
+`true` if the object is an error object.
+
+Example successful `POST /lsp/jitchannel` response:
+
+```JSON
+{
+    "jit_channel_scid": "1x4815x29451",
+    "lsp_node_id": "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f",
+    "lsp_cltv_expiry_delta" : 144,
+    "lsp_connection_strings": [
+        "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f@3.33.236.230:9735"
+    ]
+}
+```
+
+`jit_channel_scid` is the SCID to use for creating the routehint in
+the invoice, `lsp_node_id` is the node for that routehint hop, and
+`lsp_cltv_expiry_delta` is the CLTV delta for that routehint hop.
+
+`lsp_connection_strings` is how the client should connect to the
+LSP.
+
+### 3.  Invoice Generation
+
+The client then generates a [BOLT11][] invoice.
```

Comment:
Has anyone evaluated whether BOLT 12 invoices will work with this proposal?

Reply:
BOLT 12 needs onion messages.  If the client has 0 channels, it cannot be contacted via onion message.

Reply:
Hmm, I guess that is valid.  However the BOLT12 spec is intended to give a stable string, and for the initial channel opening, the plan is that the SCID is a temporary one ***only*** for the channel-opening, and after the initial payment it should be an SCID that is the  `alias` and  eventually the stable confirmed SCID.  Hmmm.

Reply:
Why is a temporary SCID needed? Can't the SCID alias be used from the start?

Reply:
The temporary SCID is used by the LSP to recognize the htlcs sent for the payment which triggers the channel creation. This is important because the client allows the LSP to substract the opening_fee from the total amount only for this specific payment.

Reply:
I am not sure if all implementations can easily have `jit_channel_scid` be the same as the `alias` SCID.  My (possibly wrong and fuzzy) recollection of the LDK API suggests that the two would be different, and I think for a CLN plugin-based solution it would be easier to implement with both being different (honestly I am not sure if 0-conf channel initiation is possible with CLN).

Reply:
+1 for different scids. Trying to push our scid to other implementations sounds like opening a huge (unnecessary) can of worms with very limited benefits.